src/Core/Grid/Definition/Factory/LinkBlockDefinitionFactory.php

```diff
@@ -0,0 +1,135 @@
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/AFL-3.0 Academic Free License 3.0 (AFL-3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\Module\LinkList\Core\Grid\Definition\Factory;
+
+use PrestaShop\PrestaShop\Core\Grid\Action\Row\RowActionCollection;
+use PrestaShop\PrestaShop\Core\Grid\Action\Row\Type\LinkRowAction;
+use PrestaShop\PrestaShop\Core\Grid\Action\Row\Type\SubmitRowAction;
+use PrestaShop\PrestaShop\Core\Grid\Column\ColumnCollection;
+use PrestaShop\PrestaShop\Core\Grid\Column\Type\Common\PositionColumn;
+use PrestaShop\PrestaShop\Core\Grid\Column\Type\Common\ActionColumn;
+use PrestaShop\PrestaShop\Core\Grid\Column\Type\DataColumn;
+use PrestaShop\PrestaShop\Core\Grid\Definition\Factory\AbstractGridDefinitionFactory;
+
+/**
+ * Class LinkBlockDefinitionFactory.
+ */
+final class LinkBlockDefinitionFactory extends AbstractGridDefinitionFactory
+{
+    const FACTORY_ID = 'link_widget_grid_';
+
+    /**
+     * @var array
+     */
+    private $hook;
+
+    /**
+     * LinkBlockDefinitionFactory constructor.
+     *
+     * @param array $hook
+     */
+    public function __construct(array $hook)
+    {
+        $this->hook = $hook;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getId()
+    {
+        return self::FACTORY_ID . $this->hook['id_hook'];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getName()
+    {
+        return $this->hook['name'] . ' ' . $this->hook['title'];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getColumns()
+    {
+        return (new ColumnCollection())
+            ->add((new DataColumn('id_link_block'))
+                ->setName($this->trans('ID', [], 'Modules.Linklist.Admin'))
+                ->setOptions([
+                    'field' => 'id_link_block',
+                ])
+            )
+            ->add((new DataColumn('block_name'))
+                ->setName($this->trans('Name of the block', [], 'Modules.Linklist.Admin'))
+                ->setOptions([
+                    'field' => 'block_name',
+                ])
+            )
+            ->add((new PositionColumn('position'))
+                ->setName($this->trans('Position', [], 'Admin.Global'))
+                ->setOptions([
+                    'field' => 'position',
+                    'id_field' => 'id_link_block',
+                    'id_parent_field' => 'id_hook',
+                    'update_route' => 'admin_link_block_update_positions',
+                    'update_method' => 'POST',
```

Comment:
Hi @jolelievre - I have some improvements regarding `PositionColumn()` functionality and how it renders the routes.

Let me define  my scenario:

- I have a controller which always has route parameter `{cmsCategoryParentId}`. It also has child components which are displayed on each row called `cmsCategoryId`.

If I want to use the `PositionColumn` I must not use any of these while generating the route because the route is generated without having in mind that it can have additional parameters. Also the `PositionColumn` would be hardly used in the form page there some kind of _routeParameter_ is mandatory.

Possible solutions:
We can have something like this when defining `PositionColumn`

```php
            ->add((new PositionColumn('position'))
                ->setName($this->trans('Position', [], 'Admin.Global'))
                ->setOptions([
                    'field' => 'position',
                    'update_method' => 'POST',
                    'update_route' => 'admin_cms_page_update_position_cms_category',
                    'route_params_by_record' => [
                        'cmsCategoryId' => 'id_cms_category',
                        'cmsCategoryParentId' => 'id_parent',
                    ],
                ])
            )
```
and in twig

```twig
  {% set routeParameters = [] %}

  {% if column.options.route_params_by_record is not empty %}
    {% for routeName, routeRecordName in column.options.route_params_by_record %}
      {% set routeRecord = record[routeRecordName] %}
      {% set routeParameters = routeParameters|merge({(routeName) : routeRecord}) %}
    {% endfor %}
  {% endif %}

  <div class="js-drag-handle js-{{ grid.id }}-position"
       style="cursor: move;"
       data-position="{{ record[column.options.field] }}"
       data-update-url="{{ url(column.options.update_route, routeParameters) }}"
       data-update-method="{{ column.options.update_method }}"
       data-pagination-offset="{{ grid.pagination.offset }}"
  >
    <i class="material-icons">swap_vert</i>
    <span class="js-position">
      {{ record[column.options.field] + 1 }}
    </span>
  </div>
```
and after this we will not need to pass them there : https://github.com/PrestaShop/PrestaShop/blob/1.7.5.x/admin-dev/themes/new-theme/js/components/grid/extension/position-extension.js#L73

and finally, the implementation

```php
    public function updateCmsCategoryPositionAction($cmsCategoryParentId, $cmsCategoryId)
    {
        // todo : implement
    }
``` 
with these changes I can pass any variables I want to the route without interacting with the core logic of what happends on javascript side.

Let me know what you think about it  :slightly_smiling_face:

Reply:
hi @tomas862 you're right I think your solution allows more use cases, although
I would generalize/simplify it a little bit more and get rid of `route_param_name/route_param_field`:

```php
         ->add((new PositionColumn('position'))
                ->setName($this->trans('Position', [], 'Admin.Global'))
                ->setOptions([
                    'field' => 'position',
                    'update_method' => 'POST',
                    'route_param_name' => 'cmsCategoryId',
                    'route_param_field' => 'id_cms_category',
                    'update_route' => 'admin_cms_page_update_position_cms_category',
                    'route_params_extra_by_record' => [
                        'cmsCategoryParentId' => 'id_parent',
                    ],
                ])
            )
```

would become:

```php
         ->add((new PositionColumn('position'))
                ->setName($this->trans('Position', [], 'Admin.Global'))
                ->setOptions([
                    'field' => 'position',
                    'update_method' => 'POST',
                    'update_route' => 'admin_cms_page_update_position_cms_category',
                    'route_params' => [
                        //Writing it this way seems more logical to me since
                        //id_cms_category is the available field in the and
                        //cmsCategoryId is what we try to fill
                        'id_cms_category' => 'cmsCategoryId',
                        'id_parent' => 'cmsCategoryParentId',
                    ],
                ])
            )
```

Reply:
And for the question about allow multiple router params, it's logic and not over enginerring to allow multiple params.

Reply:
ok so I am going ton integrate this feature @tomas862 it should be available in 1.7.5 pretty soon, so you will be able to use it for cms pages

Reply:
Thanks @jolelievre  and @PierreRambaud  :+1: